flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/lookup/RetryableAsyncLookupFunctionDelegator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.operators.join.lookup;
+
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.functions.AsyncLookupFunction;
+import org.apache.flink.table.functions.FunctionContext;
+
+import javax.annotation.Nonnull;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Predicate;
+
+/** A delegator holds user's {@link AsyncLookupFunction} to handle retries. */
+public class RetryableAsyncLookupFunctionDelegator extends AsyncLookupFunction {
+
+    private final AsyncLookupFunction userLookupFunction;
+
+    private final ResultRetryStrategy retryStrategy;
+
+    private final boolean retryEnabled;
+
+    private transient Predicate<Collection<RowData>> retryResultPredicate;
+
+    public RetryableAsyncLookupFunctionDelegator(
+            @Nonnull AsyncLookupFunction userLookupFunction,
+            @Nonnull ResultRetryStrategy retryStrategy) {
+        this.userLookupFunction = userLookupFunction;
+        this.retryStrategy = retryStrategy;
```

Comment:
please do some null check for given arguments

Reply:
will add checkNotNull